Completion glyphs ( Bump to version 0.15.0)

Krzysztof-Cieslak opened the pull request:
As discussed in https://github.com/fsharp/fsharpbinding/issues/1010 :
- I've updated Completion API to get also glyph number ( we can consider translating it too meaningful string)
- I've committed updated test results.

Also: 
- I've fixed `paket.references` file naming in `FSharp.AutoComplete` project
- I've removed `build.fsx` reference from `FSharp.AutoComplete.fsproj`
- I've added `build.fsx`, `paket.lock` and `readme.md` files to `sln` (I think it's standard way of managing solutions with Paket and FAKE)

Krzysztof-Cieslak committed:
Add Glyph to Completion output

rneatherway commented:
Thanks! The naming of the paket.references file is actually deliberate to prevent it from affecting the project files in the tests folder (http://fsprojects.github.io/Paket/references-files.html#File-name-conventions).

When you say we could consider adding the conversion to string, would that be in a later PR? It would be preferable not to break the API twice if possible.

Krzysztof-Cieslak commented:
I've reversed name change in `paket.references` file - i was bit confused since `fsproj` file referenced not existing `paket.references` instead of `FSharp.AutoComplete.paket.references`

Here is my function translating Glyph number to string (based of Dave's list, but I had to make it more FunScript friendly) used in Atom plugin right now - https://github.com/fsprojects/FSharp.Atom/blob/develop/src/Autocomplete.fs#L30. We can just move it here if You think it's better choice. Or we can move function Dave linked to just provide better string names. Your call.

7sharp9 commented:
I think I only use the glyph list in one place now, the pathed document drop down:
![screen shot 2015-05-18 at 09 05 27](https://cloud.githubusercontent.com/assets/588746/7676585/12f1888e-fd3d-11e4-9ba2-30d4c46af742.png)

Its been depreciated by the new symbol API functions, which use this:

![screen shot 2015-05-18 at 09 06 58](https://cloud.githubusercontent.com/assets/588746/7676611/472da0d8-fd3d-11e4-9dce-4293a8053e1d.png)

Krzysztof-Cieslak commented:
@7sharp9: I want to use glyphs for getting completion type in autocomplete - here is current version of it ![pasted_image_at_2015_05_17_17_57](https://cloud.githubusercontent.com/assets/5427083/7677243/b07f57e0-fd4a-11e4-96f8-c306fc2c7b87.png)

kjnilsson commented:
looking good - is there a canonical (or something we can all agree on) single-letter representation of these glyphs? if so it would be good to include that in the return object as well as the full name.

Krzysztof-Cieslak commented:
For almost all we can just use starting letter of word. Problems: 
- "Enum", "Exception", "Extension Method"
- "Method", "Module"

rneatherway commented:
@Krzysztof-Cieslak yes I forgot to update FSharp.AutoComplete.fsproj with the new paket references filename.

@7sharp9 so we should be calling `GetDeclarationSymbols` now instead of `GetDeclarations` (https://github.com/fsharp/FSharp.AutoComplete/blob/master/FSharp.AutoComplete/Program.fs#L472)? I took a look at the FCS code and even that is now marked as deprecated: https://github.com/fsharp/FSharp.Compiler.Service/blob/master/src/fsharp/vs/service.fs#L1918

It looks like it has been replaced by `GetDeclarationListSymbols`?

For the one-letter, some clashes are survivable probably. 'x' is OK for exception I think, if we only want to use one case then it is hard to distinguish method and module though.

rneatherway commented:
@7sharp9 ok I found that it is a wrapper in FSharp.CompilerBinding, but why is it a list of lists now rather than just a list?

7sharp9 commented:
Why is what a list of lists?  The declarations?  For overloads, the list of overloads is also present in the non symbol one albeit wrapped in the declaration wrapper type.

rneatherway commented:
`GetDeclarations` just returns a list, but yes, `GetDeclarationSymbols` returns a list of lists. Is that because each member of the list is a list of overloads with the same name?

rneatherway commented:
OK, I've had more of an explore, and the new API looks interesting, but it doesn't give anything more we need right now, so let's just add this feature. @Krzysztof-Cieslak, could you:
1. Rebase on master (I fixed the paket.references issue, which needs to have fsproj in the name according to the Paket docs).
2. Add the conversion function to FSharp.AutoComplete, with a single character option (GlyphChar?) as well. I think use 'X' for Exception, 'N' for Module (like namespace) and 'M' for Extension Method.
3. Don't output the glyph in text mode, that is just for testing really (it's too hard to check by hand) and we have a JSON test as well.

Then I'll test and merge, thanks!

7sharp9 commented:
@rneatherway Unless your using symbols as part of navigation and searching then they wont be as useful.  In XS we use the symbol to format the tooltips and autocompletions rather then using lots of string cutting/parsing operations to format for type colour etc.  ToolTipText types were a real pain to work with.

rneatherway commented:
Yeah I was having a look over your code and it looks much easier to work with for doing that kind of formatting. Things over here are still a bit more basic, but hopefully we will migrate over to using the symbols in due course. It's hard to know exactly what to do because the display features of the various editors are pretty different.

Krzysztof-Cieslak committed:
Translating glyph to string and glyph character

Krzysztof-Cieslak committed:
Update test results

Krzysztof-Cieslak commented:
OK, so I've rebased PR and have updated implementation with translating  `int` to meaningful `string` ( basing on function Dave provided). I've added GlyphChar to JSON result of completion command. 

Also I've updated JSON test results to match new Completion API

rneatherway commented on FSharp.AutoComplete/Program.fs:
```diff
@@ -474,7 +523,7 @@ module internal Main =
                   | Text ->
                       printAgent.WriteLine "DATA: completion"
-                      for d in decls.Items do printAgent.WriteLine(d.Name)
+                      for d in decls.Items do printAgent.WriteLine(sprintf "%s" d.Name)
```
Could you revert this line please?

rneatherway commented on FSharp.AutoComplete/Program.fs:
```diff
@@ -267,6 +274,48 @@ module internal CommandInput =
+      0x0008,  ("Interface", "I")
+      0x000e,  ("Module", "N") (* module *)
+      0x000f,  ("Namespace", "N")
+      0x000c,  ("Method", "m")
```
Can we use upper-case for consistency?

rneatherway replied:
I mean 'M'

rneatherway commented:
Thanks, two small comments, then I'll merge and do a release tomorrow.

Krzysztof-Cieslak committed:
Small fixes

Krzysztof-Cieslak commented:
Done.

rneatherway commented:
Great, the release is now available at https://github.com/fsharp/FSharp.AutoComplete/releases/tag/0.15.0
